geomstats/geometry/sasaki_metric.py

```diff
@@ -0,0 +1,238 @@
+            Tangent vector at the base point equal to the Riemannian logarithm
+            of point at the base point.
+        """
+        # unflatten
+        pts = gs.reshape(point, (-1, 2) + self.metric.shape)
+        bs_pts = gs.reshape(base_point, (-1, 2) + self.metric.shape)
+
+        metric = self.metric
+        par_trans = metric.parallel_transport
+        n_s = self.n_s if n_steps is None else n_steps
+
+        @delayed
+        def do_log(pt, bs_pt):
+            """Calculate the logarithm."""
+            pu = self.geodesic_discrete(bs_pt, pt, n_s)
+            p1, u1 = pu[1][0], pu[1][1]
+            p0, u0 = bs_pt[0], bs_pt[1]
+            w = (par_trans(u1, p1, None, p0) - u0)
+            v = metric.log(point=p1, base_point=p0)
+            return n_s * gs.array([v, w])
+
+        n_jobs = min(os.cpu_count(), len(pts))
+        with Parallel(n_jobs=n_jobs, verbose=0) as parallel:
+            rslt = parallel(do_log(pt, bs_pts[i % len(bs_pts)])
+                            for i, pt in enumerate(pts))
+
+        return gs.reshape(gs.array(rslt), point.shape)
+
+    def geodesic_discrete(self, initial_point, end_point, n_steps=None, **kwargs):
+        """Compute Sakai geodesic employing a variational time discretization.
+
+        Parameters
+        ----------
+        end_point : array-like, shape=[2, M.shape]
+            Point in the tangent bundle TM of manifold M.
+        initial_point : array-like, shape=[2, M.shape]
+            Point in the tangent bundle TM of manifold M.
+        n_steps : int
+            n_steps - 1 is the number of intermediate points in the discretization
+            of the geodesic from initial_point to end_point
+            Optional, default: self.n_s.
+
+        Returns
+        -------
+        geodesic : array-like, shape=[n_steps + 1, 2, M.shape]
+            Discrete geodesic x(s)=(p(s), u(s)) in Sasaki metric connecting
+            initial_point = x(0) and end_point = x(1).
+        """
+        n_s = int(self.n_s if n_steps is None else n_steps)
+        metric = self.metric
+        par_trans = metric.parallel_transport
+        p0, u0 = initial_point[0], initial_point[1]
+        pL, uL = end_point[0], end_point[1]
+
+        def grad(pu):
+            """Gradient of discrete geodesic energy."""
+            g = []
+            # add boundary points to the list of points
+            pu = [initial_point] + pu + [end_point]
+            for j in range(n_s - 1):
+                p1, _ = pu[j][0], pu[j][1]
+                p2, u2 = pu[j + 1][0], pu[j + 1][1]
+                p3, u3 = pu[j + 2][0], pu[j + 2][1]
+                v, w = metric.log(p3, p2), par_trans(u3, p3, None, p2) - u2
+                gp = metric.log(p3, p2) + metric.log(p1, p2) + metric.curvature(
+                    u2, w, v, p2)
+                gu = par_trans(u3, p3, None, p2) - 2 * u2 + par_trans(u0, p0, None, p2)
+                g.append([gp, gu])
+            return -n_s * gs.array(g)
```

Comment:
This can be vectorized:

```
def grad(pu):
  pu = gs.vstack([gs.expand_dims(initial_point, axis=0),
            pu_ini,
            gs.expand_dims(end_point, axis=0)])

  p = gs.take(pu, 0, axis=1)
  u = gs.take(pu, 1, axis-1)

  p1, p2, p3 = p[:-2], p[1:-1], p[2:]
  u2, u3 = u[1:-1], u[2:]

  v = metric.log(p3, p2)
  w = par_trans(u3, p3, end_point=p2) - u2

  gp = metric.log(p3, p2) + metric.log(p1, p2) + metric.curvature(
      u2, w, v, p2)
  
  gu = par_trans(u3, p3, end_point=p2) - 2 * u2 + par_trans(u0, p0, end_point=p2)

  return -n_s*gs.stack([gp, gu], axis=1)
```

The expressions are slightly different from the ones in the paper. Do you have any reference with these expressions?

Reply:
The formulas used here can all be juggled out of the ones in the paper by using definitions and basic properties.  However, and here we are really thankful for pointing on this, there was introduced some mistake when porting the code form Matlab to Python.  I checked back with Esfandiar and @martinhanik and put in some corrections.

Reply:
Do you have any notes with such manipulations so we can point our users to them in the documentation?

Reply:
We created a little tech-report with detailed explainations (https://nbn-resolving.org/urn/resolver.pl?urn:nbn:de:0297-zib-87174) and reference it now in the docsting of the sasaki class.